lib/nvml.ml

```diff
@@ -136,13 +138,59 @@ let get_vgpu_for_uuid iface vgpu_uuid vgpus =
     )
     vgpus
 
-(* mock implementation *)
-module NVML = struct
-  let attach () = ()
+module NVML : sig
+  val attach : unit -> unit
+
+  val detach : unit -> unit
+
+  val is_attached : unit -> bool
+
+  val get : unit -> interface option
+end = struct
+  let interface = ref (None : interface option)
+
+  let mx = Mutex.create ()
 
-  let detach () = ()
+  let finally = Xapi_stdext_pervasives.Pervasiveext.finally
 
-  let is_attached () = true
+  let with_mutex mx f =
+    Mutex.lock mx ;
+    finally f (fun () -> Mutex.unlock mx)
+
+  let open_nvml_interface () =
+    let interface = library_open () in
+    try init interface ; interface with e -> library_close interface ; raise e
+
+  let get () = !interface
```

Comment:
I wonder if it would be wiser to have

```ocaml
val with_nvml : (interface option -> unit) -> unit
```
That can holds and drops the mutex related to the interface handle to avoid race conditions

Reply:
Currently the mutex is only held while updating the value that holds the interface. Do you suggest to hold it while the interface/library is loaded?

Reply:
I meant holding the mutex when the interface is being read, would that cause problems because there's more than one thread trying to read it?